x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -2,46 +2,123 @@ package math
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	if kAtPriceOne.GTE(sdk.ZeroInt()) {
+		kIncrementDistance = sdk.NewDec(9).Mul(sdk.OneDec().Quo(sdk.NewInt(10).ToDec().Power(kAtPriceOne.Uint64())))
 	} else {
-		sqrtPrice, err = sdk.OneDec().Quo(sdk.NewDecWithPrec(10001, 4).Power(tickIndex.Abs().Uint64())).ApproxSqrt()
+		kIncrementDistance = sdk.NewDec(9).Mul(sdk.NewInt(10).ToDec().Power(kAtPriceOne.Abs().Uint64()))
 	}
-	if err != nil {
-		return sdk.Dec{}, err
+
+	// Use floor division to determine how many k increments we have passed
+	kDelta := tickIndex.ToDec().Quo(kIncrementDistance).TruncateInt()
+
+	// Calculate the current k value from the starting k value and the k delta
+	curK := (kAtPriceOne.Add(kDelta))
+
+	var curIncrement sdk.Dec
+	if curK.GTE(sdk.ZeroInt()) {
+		curIncrement = sdk.NewDec(10).Power(curK.Uint64())
+	} else {
+		curIncrement = sdk.NewDec(1).Quo(sdk.NewDec(10).Power(curK.Abs().Uint64()))
 	}
 
-	return sqrtPrice, nil
+	numAdditiveTicks := tickIndex.ToDec().Sub(kDelta.ToDec().Mul(kIncrementDistance))
+
+	if kDelta.GTE(sdk.ZeroInt()) {
+		price = sdk.NewDec(10).Power(kDelta.Uint64()).Add(numAdditiveTicks.Mul(curIncrement))
+	} else {
+		price = sdk.OneDec().Quo(sdk.NewDec(10).Power(kDelta.Abs().Uint64())).Add(numAdditiveTicks.Mul(curIncrement))
+	}
+	return price, nil
 }
 
 // PriceToTick takes a price and returns the corresponding tick index
-func PriceToTick(price sdk.Dec) sdk.Int {
-	tick := osmomath.BigDecFromSDKDec(price).CustomBaseLog(osmomath.NewDecWithPrec(10001, 4))
-	return tick.SDKDec().TruncateInt()
+func PriceToTick(price sdk.Dec, kAtPriceOne sdk.Int) (tickIndex sdk.Int) {
+	if price.Equal(sdk.OneDec()) {
+		return sdk.ZeroInt()
+	}
+
+	var kIncrementDistance sdk.Dec
+	// The formula is as follows: k_increment_distance = 9 * 10**(-k_at_price_1)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	if kAtPriceOne.GTE(sdk.ZeroInt()) {
+		kIncrementDistance = sdk.NewDec(9).Mul(sdk.OneDec().Quo(sdk.NewInt(10).ToDec().Power(kAtPriceOne.Uint64())))
+	} else {
+		kIncrementDistance = sdk.NewDec(9).Mul(sdk.NewInt(10).ToDec().Power(kAtPriceOne.Abs().Uint64()))
+	}
+
+	total := sdk.OneDec()
+	ticksPassed := sdk.ZeroInt()
+	currentK := kAtPriceOne
+
+	var curIncrement sdk.Dec
+	if currentK.GTE(sdk.ZeroInt()) {
+		curIncrement = sdk.NewDec(10).Power(currentK.Uint64())
+	} else {
+		curIncrement = sdk.NewDec(1).Quo(sdk.NewDec(10).Power(currentK.Abs().Uint64()))
+	}
+
+	for total.LT(price) {
+		if currentK.GTE(sdk.ZeroInt()) {
+			curIncrement = sdk.NewDec(10).Power(currentK.Uint64())
+			maxPriceForCurrentIncrement := kIncrementDistance.Mul(curIncrement)
+			if total.Add(maxPriceForCurrentIncrement).LT(price) {
+				total = total.Add(maxPriceForCurrentIncrement)
+				currentK = currentK.Add(sdk.OneInt())
+				ticksPassed = ticksPassed.Add(kIncrementDistance.TruncateInt())
+			} else {
+				break
+			}
```

Comment:
all this logic can be shared if moved outside of the surrounding if / else

Reply:
I think I addressed what you were referring to